<a href="https://colab.research.google.com/github/julienessan/Gomycode_Cours_Data_Sciences/blob/main/Checkpoint_Data_pre_processing_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Checkpoint: Exploration des données avec Pandas**

**Objectif:**

Dans ce point de contrôle, nous visons à explorer et à prétraiter l'historique de facturation des clients de la compagnie tunisienne d'électricité et de gaz de 2005 à 2019, en nous concentrant sur la compréhension de la structure des données, le traitement des valeurs manquantes, la réalisation d'une analyse descriptive et la transformation des variables catégorielles en formats numériques.

In [2]:
import pandas as pd
import numpy as np

In [3]:
client_0_bills = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/STEG_BILLING_HISTORY.csv")
client_0_bills.head(10)

/tmp/ipykernel_724/3310835241.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  client_0_bills = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/STEG_BILLING_HISTORY.csv")


,client_id,invoice_date,tarif_type,counter_number,counter_statue,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number,counter_type
0,train_Client_0,2014-03-24,11,1335667.0,0,203,8.0,1,82,0,0,0,14302,14384,4,ELEC
1,train_Client_0,2013-03-29,11,1335667.0,0,203,6.0,1,1200,184,0,0,12294,13678,4,ELEC
2,train_Client_0,2015-03-23,11,1335667.0,0,203,8.0,1,123,0,0,0,14624,14747,4,ELEC
3,train_Client_0,2015-07-13,11,1335667.0,0,207,8.0,1,102,0,0,0,14747,14849,4,ELEC
4,train_Client_0,2016-11-17,11,1335667.0,0,207,9.0,1,572,0,0,0,15066,15638,12,ELEC
5,train_Client_0,2017-07-17,11,1335667.0,0,207,9.0,1,314,0,0,0,15638,15952,8,ELEC
6,train_Client_0,2018-12-07,11,1335667.0,0,207,9.0,1,541,0,0,0,15952,16493,12,ELEC
7,train_Client_0,2019-03-19,11,1335667.0,0,207,9.0,1,585,0,0,0,16493,17078,8,ELEC
8,train_Client_0,2011-07-22,11,1335667.0,0,203,9.0,1,1200,186,0,0,7770,9156,4,ELEC
9,train_Client_0,2011-11-22,11,1335667.0,0,203,6.0,1,1082,0,0,0,9156,10238,4,ELEC


In [3]:
# type de données de la variable 'client_0_bills
print(type(client_0_bills))

<class 'pandas.core.frame.DataFrame'>


In [4]:
# les informations générales de l'ensemble de données
client_0_bills.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4476749 entries, 0 to 4476748
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   int64  
 8   consommation_level_1  int64  
 9   consommation_level_2  int64  
 10  consommation_level_3  int64  
 11  consommation_level_4  int64  
 12  old_index             int64  
 13  new_index             int64  
 14  months_number         int64  
 15  counter_type          object 
dtypes: float64(2), int64(10), object(4)
memory usage: 546.5+ MB


Combien de lignes et de colonnes avons-nous dans cet ensemble de données ?

**Lignes: 4476749**

**Colonne: 16**


Combien de caractéristiques catégorielles sont présentes dans l'ensemble de données ?

**caractéristiques catégorielles = 4**


Combien d'espace mémoire l'ensemble de données consomme-t-il ?

memory usage= 546.5+ MB



In [5]:
# Inspectez l'ensemble de données pour détecter d'éventuelles valeurs manquantes.
client_0_bills.isnull().sum()

,0
client_id,0
invoice_date,0
tarif_type,0
counter_number,48
counter_statue,0
counter_code,0
reading_remarque,4531
counter_coefficient,0
consommation_level_1,0
consommation_level_2,0


### Stratégie de traitement des valeurs manquantes

**1. Pour la colonne `counter_number`:**
Étant donné que le nombre de valeurs manquantes (48) est très faible par rapport à la taille totale du dataset, nous allons **supprimer les lignes** contenant ces valeurs manquantes. Cela minimisera la perte de données et évitera d'introduire des biais par imputation.

**2. Pour la colonne `reading_remarque`:**
Cette colonne présente 4531 valeurs manquantes. Comme elle semble être une variable de type catégoriel (indiquant une 'remarque de lecture'), nous allons **imputer les valeurs manquantes par le mode** (la valeur la plus fréquente) de cette colonne. Cela permettra de conserver les données restantes tout en remplissant les lacunes de manière statistiquement pertinente.

In [6]:
# 1. Traitement de 'counter_number' : suppression des lignes avec des valeurs manquantes
client_0_bills.dropna(subset=['counter_number'], inplace=True)

# 2. Traitement de 'reading_remarque' : imputation par le mode
mode_reading_remarque = client_0_bills['reading_remarque'].mode()[0]
client_0_bills['reading_remarque'] = client_0_bills['reading_remarque'].fillna(mode_reading_remarque)

# Vérification après traitement
print("Vérification des valeurs manquantes après traitement :")
client_0_bills.isnull().sum()

Vérification des valeurs manquantes après traitement :


,0
client_id,0
invoice_date,0
tarif_type,0
counter_number,0
counter_statue,0
counter_code,0
reading_remarque,0
counter_coefficient,0
consommation_level_1,0
consommation_level_2,0


In [7]:
# analyse descriptive sur les caractéristiques numériques
client_0_bills.describe()

,tarif_type,counter_number,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number
count,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06,4.476701e+06
mean,2.012810e+01,1.230595e+11,1.724879e+02,7.320377e+00,1.003040e+00,4.109775e+02,1.093204e+02,2.030547e+01,5.292247e+01,1.776699e+04,1.834969e+04,4.483138e+01
std,1.347258e+01,1.657276e+12,1.338872e+02,1.571627e+00,3.083483e-01,7.573100e+02,1.220126e+03,1.574199e+02,8.754571e+02,4.036702e+04,4.095329e+04,3.128352e+03
min,8.000000e+00,0.000000e+00,0.000000e+00,5.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,1.100000e+01,1.211080e+05,5.000000e+00,6.000000e+00,1.000000e+00,7.900000e+01,0.000000e+00,0.000000e+00,0.000000e+00,1.791000e+03,2.056000e+03,4.000000e+00
50%,1.100000e+01,4.945610e+05,2.030000e+02,8.000000e+00,1.000000e+00,2.740000e+02,0.000000e+00,0.000000e+00,0.000000e+00,7.690000e+03,8.192000e+03,4.000000e+00
75%,4.000000e+01,1.115161e+06,2.070000e+02,9.000000e+00,1.000000e+00,6.000000e+02,0.000000e+00,0.000000e+00,0.000000e+00,2.166000e+04,2.234300e+04,4.000000e+00
max,4.500000e+01,2.798115e+13,6.000000e+02,4.130000e+02,5.000000e+01,9.999100e+05,9.990730e+05,6.449200e+04,5.479460e+05,2.800280e+06,2.870972e+06,6.366240e+05


In [8]:
# Sélectionnez les enregistrements de factures pour le client avec un id = 'train_Client_0', en utilisant 2 méthodes.

#1er methode
selected_train_Client_0 = client_0_bills[client_0_bills['client_id'] == 'train_Client_0']
selected_train_Client_0.head()


,client_id,invoice_date,tarif_type,counter_number,counter_statue,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number,counter_type
0,train_Client_0,2014-03-24,11,1335667.0,0,203,8.0,1,82,0,0,0,14302,14384,4,ELEC
1,train_Client_0,2013-03-29,11,1335667.0,0,203,6.0,1,1200,184,0,0,12294,13678,4,ELEC
2,train_Client_0,2015-03-23,11,1335667.0,0,203,8.0,1,123,0,0,0,14624,14747,4,ELEC
3,train_Client_0,2015-07-13,11,1335667.0,0,207,8.0,1,102,0,0,0,14747,14849,4,ELEC
4,train_Client_0,2016-11-17,11,1335667.0,0,207,9.0,1,572,0,0,0,15066,15638,12,ELEC


In [9]:
# 2e methode avec loc:
selected_train_Client_0_2 = client_0_bills.loc[client_0_bills['client_id'] == 'train_Client_0']
selected_train_Client_0_2.head()

,client_id,invoice_date,tarif_type,counter_number,counter_statue,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number,counter_type
0,train_Client_0,2014-03-24,11,1335667.0,0,203,8.0,1,82,0,0,0,14302,14384,4,ELEC
1,train_Client_0,2013-03-29,11,1335667.0,0,203,6.0,1,1200,184,0,0,12294,13678,4,ELEC
2,train_Client_0,2015-03-23,11,1335667.0,0,203,8.0,1,123,0,0,0,14624,14747,4,ELEC
3,train_Client_0,2015-07-13,11,1335667.0,0,207,8.0,1,102,0,0,0,14747,14849,4,ELEC
4,train_Client_0,2016-11-17,11,1335667.0,0,207,9.0,1,572,0,0,0,15066,15638,12,ELEC


### Conversion des types de données

Transformez la caractéristique 'counter_type' en une variable numérique en utilisant l'encodeur de votre choix.

In [4]:
# Transformez la caractéristique 'counter_type' en une variable numérique en utilisant l'encodeur de votre choix.
client_0_bills['counter_type'] = client_0_bills['counter_type'].astype('category').cat.codes
# Vérification des types de données après conversion
client_0_bills

,client_id,invoice_date,tarif_type,counter_number,counter_statue,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number,counter_type
0,train_Client_0,2014-03-24,11,1335667.0,0,203,8.0,1,82,0,0,0,14302,14384,4,0
1,train_Client_0,2013-03-29,11,1335667.0,0,203,6.0,1,1200,184,0,0,12294,13678,4,0
2,train_Client_0,2015-03-23,11,1335667.0,0,203,8.0,1,123,0,0,0,14624,14747,4,0
3,train_Client_0,2015-07-13,11,1335667.0,0,207,8.0,1,102,0,0,0,14747,14849,4,0
4,train_Client_0,2016-11-17,11,1335667.0,0,207,9.0,1,572,0,0,0,15066,15638,12,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4476744,train_Client_99998,2005-08-19,10,1253571.0,0,202,9.0,1,400,135,0,0,3197,3732,8,0
4476745,train_Client_99998,2005-12-19,10,1253571.0,0,202,6.0,1,200,6,0,0,3732,3938,4,0
4476746,train_Client_99999,1996-09-25,11,560948.0,0,203,6.0,1,259,0,0,0,13884,14143,4,0
4476747,train_Client_99999,1996-05-28,11,560948.0,0,203,6.0,1,603,0,0,0,13281,13884,4,0


In [5]:
# Supprimez la caractéristique 'counter_statue' du Dataframe
client_0_bills.drop(columns=['counter_statue'], inplace=True)
client_0_bills

,client_id,invoice_date,tarif_type,counter_number,counter_code,reading_remarque,counter_coefficient,consommation_level_1,consommation_level_2,consommation_level_3,consommation_level_4,old_index,new_index,months_number,counter_type
0,train_Client_0,2014-03-24,11,1335667.0,203,8.0,1,82,0,0,0,14302,14384,4,0
1,train_Client_0,2013-03-29,11,1335667.0,203,6.0,1,1200,184,0,0,12294,13678,4,0
2,train_Client_0,2015-03-23,11,1335667.0,203,8.0,1,123,0,0,0,14624,14747,4,0
3,train_Client_0,2015-07-13,11,1335667.0,207,8.0,1,102,0,0,0,14747,14849,4,0
4,train_Client_0,2016-11-17,11,1335667.0,207,9.0,1,572,0,0,0,15066,15638,12,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4476744,train_Client_99998,2005-08-19,10,1253571.0,202,9.0,1,400,135,0,0,3197,3732,8,0
4476745,train_Client_99998,2005-12-19,10,1253571.0,202,6.0,1,200,6,0,0,3732,3938,4,0
4476746,train_Client_99999,1996-09-25,11,560948.0,203,6.0,1,259,0,0,0,13884,14143,4,0
4476747,train_Client_99999,1996-05-28,11,560948.0,203,6.0,1,603,0,0,0,13281,13884,4,0
